src/commands/supported-tokens.ts

```diff
@@ -0,0 +1,287 @@
+   * This chunk size balances between:
+   * - RPC rate limits (avoid too many parallel requests)
+   * - Performance (process multiple tokens simultaneously)
+   * - Memory usage (keep Promise.all() batches reasonable)
+   * Increase for faster processing if your RPC can handle it.
+   * Decrease if you encounter rate limits or timeouts.
+   */
+  PARALLEL_POOL_CHECKS: 5,
+} as const
+
+export async function showSupportedTokens(
+  providers: Providers,
+  argv: { source: string; router: string; dest: string; format: Format },
+) {
+  console.log('[INFO] Starting token discovery for cross-chain transfers')
+  const sourceChainId = isNaN(+argv.source) ? chainIdFromName(argv.source) : +argv.source
+  const sourceProvider = await providers.forChainId(sourceChainId)
+
+  const destChainId = isNaN(+argv.dest) ? chainIdFromName(argv.dest) : +argv.dest
+  const destSelector = chainSelectorFromId(destChainId)
+
+  const router = new Contract(argv.router, RouterABI, sourceProvider) as unknown as TypedContract<
+    typeof RouterABI
+  >
+
+  const isChainSupported = await router.isChainSupported(destSelector)
+  if (!isChainSupported) {
+    throw new Error(
+      `Lane "${chainNameFromId(sourceChainId)}" -> "${chainNameFromId(destChainId)}" is not supported`,
+    )
+  }
+  console.log(
+    `[INFO] Lane "${chainNameFromId(sourceChainId)}" -> "${chainNameFromId(destChainId)}" is supported`,
+  )
+
+  // Get onRamp address from router
+  const onRampAddress = await router.getOnRamp(destSelector)
+  const [lane, onrampContract] = await getOnRampLane(sourceProvider, onRampAddress.toString())
+
+  // Get registry address from onRamp's static config
+  const staticConfig = await onrampContract.getStaticConfig()
+
+  // Get registry address based on version
+  if (lane.version === CCIPVersion_1_2) throw new Error('Deprecated CCIP onRamp version')
+  const registryAddress = (staticConfig as { tokenAdminRegistry: string | Addressable })
+    .tokenAdminRegistry
+
+  console.log(`[INFO] Using TokenAdminRegistry at ${registryAddress.toString()}`)
+  const registry = new Contract(
+    registryAddress,
+    TokenAdminRegistryABI,
+    sourceProvider,
+  ) as unknown as TypedContract<typeof TokenAdminRegistryABI>
+
+  // Handle pagination
+  let startIndex = 0n
+  const maxCount = CONFIG.BATCH_SIZE
+  const tokenToPoolMap: Record<string, string> = {}
+  let tokensBatch: Array<string | Addressable> = []
+  let totalScanned = 0
+
+  console.log(
+    `[INFO] Fetching all registered tokens from "${chainNameFromId(sourceChainId)}" using TokenAdminRegistry`,
+  )
+
+  do {
+    console.log(`[INFO] Fetching batch: offset=${startIndex}, limit=${maxCount}`)
+    tokensBatch = [...(await registry.getAllConfiguredTokens(startIndex, BigInt(maxCount)))]
+    totalScanned += tokensBatch.length
+    console.log(`[INFO] Found ${tokensBatch.length} tokens (total scanned: ${totalScanned})`)
+
+    if (tokensBatch.length > 0) {
```

Comment:
Could/should some of these "big steps" be split into local util functions (in same module, until they're needed by other modules).